streams/src/main/java/org/apache/kafka/streams/state/internals/BufferValue.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Objects;
+
+public final class BufferValue {
+    private final ContextualRecord record;
+    private final byte[] priorValue;
+
+    BufferValue(final ContextualRecord record,
+                final byte[] priorValue) {
+        this.record = record;
+        this.priorValue = priorValue;
+    }
+
+    ContextualRecord record() {
+        return record;
+    }
+
+    byte[] priorValue() {
+        return priorValue;
+    }
+
+    static BufferValue deserialize(final ByteBuffer buffer) {
+        final ContextualRecord record = ContextualRecord.deserialize(buffer);
+
+        final int priorValueLength = buffer.getInt();
+        if (priorValueLength == -1) {
+            return new BufferValue(record, null);
+        } else {
+            final byte[] priorValue = new byte[priorValueLength];
+            buffer.get(priorValue);
+            return new BufferValue(record, priorValue);
+        }
+    }
+
+    byte[] serialize() {
+        final byte[] serializedRecord = record.serialize();
```

Comment:
It's a little more complicated then that, since we don't know the size of the serialized RecordContext until we actually serialize it (everything else is already serialized). But I'll see what I can do...

Reply:
Addressed this by introducing a container class for the result that achieves the same effect of being able to distinguish between "absent" and `null`.